Support session

merav-aharoni opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Adding support for Session in `qiskit-ibm-provider`.


### Details and comments
A `_session` member was added to IBMBackend.
This update introduces duplications between this repo and qiskit-ibm-runtime. They should be addressed after this one is merged. See related issue: https://github.com/Qiskit/qiskit-ibm-runtime/issues/969.
Fixes #588.

merav-aharoni commented:
I implemented the interface as @jyu00 suggested in #588. I decided to implement `session` as a member of `IBMBackend` and not as an `option`, because the options here all come from the configuration, and I thought `session` would be too different than the other options.

merav-aharoni commented:
We also need to update the relevant tutorial. I prefer to do this in a separate PR, so that we first finalize the implementation details here.

kt474 commented on releasenotes/notes/session_support-90c41a3c3f58a46b.yaml:
```diff
@@ -0,0 +1,16 @@
+  - |
+    Added support for using session. The syntax here is slightly different than that
+    in ``qiskit-ibm-runtime``, but the functionality is similar.
+    Here is an example using session as a context manager::
+    
+      with backend.open_session() as session:
+        job = backend.run(ReferenceCircuits.bell())
+    
```
```suggestion
    Here is an example using session as a context manager:
    
    .. code-block::
    
      with backend.open_session() as session:
        job = backend.run(ReferenceCircuits.bell())
    
```
For code snippets in the release notes I think they need to be prefaced with `.. code-block::`

merav-aharoni replied:
I am fine with your suggestion. I must say I saw other examples in the release notes without `code-block`, so not sure if this is mandatory.

kt474 commented on qiskit_ibm_provider/ibm_backend.py:
```diff
@@ -529,13 +534,28 @@ def _runtime_run(
+        session = self._session
+        if session:
+            if not session.active:
+                raise RuntimeError(f"The session {session.session_id} is closed.")
```
Is this the expected behavior? What if something like the following is run, shouldn't the 2nd job run normally with no session? 

```python
backend.open_session()
job = backend.run(circ)
backend.close_session()

job2 = backend.run(circ) <-- should this job run normally, without session
```

Also I think just running `job = backend.run(circ)` should preserve the previous behavior and not use sessions at all

merav-aharoni replied:
I accept your approach. I modified the code and the tests accordingly.

kt474 commented on qiskit_ibm_provider/ibm_backend.py:
```diff
@@ -529,13 +534,28 @@ def _runtime_run(
+        else:
+            session_id = None
+            max_execution_time = None
+        start_session = session_id is None
```
I think the logic might have to be altered a little because we don't want to have `start_session` as `True` if the user doesn't actually want to start a session.

right now something like this would still start a session:
```python
job = backend.run(circ)
```

merav-aharoni replied:
> 

You are right. I fixed that.

merav-aharoni commented:
I made a few additional changes since your last review:

1. Added/modified checks inside the tests following your comment above.
2. Added handling of `session_id` in `IBM_Job`. I noticed this was missing while doing (1).
3. Added the new test file to `integration-test-3` in the Makefile.
I think this will require an additional review.
